ENH: Adds 'altitude' to test instruments

jklenzing opened the pull request:
# Description

Closes #410.

- Adds a constant altitude of 400 km to the test instruments.
- Updates `pysat_testing2d_xarray` to use `altitude_profile' and 'altitude_profile2' for the profile data.

Required for pysat/pysatModels#34.

## Type of change

- New feature (non-breaking change which adds functionality)
- Breaking change (fix or feature that would cause existing functionality
      to not work as expected)

# How Has This Been Tested?

via pytest.

Also by 
```
import pysat
test = pysat.Instrument('pysat', 'testing')
test.load(date=pysat.datetime(2009, 1, 1))
print(test)
```
which should result in 'altitude' appearing in the list of variables.  I did this for each of the four satellite objects without error.

**Test Configuration**:
* Mac OS X 10.15.4
* python 3.7.3

# Checklist:

- [x] Make sure you are merging into the ``develop`` (not ``master``) branch
- [x] My code follows the style guidelines of this project
- [x] I have performed a self-review of my own code
- [x] I have commented my code, particularly in hard-to-understand areas
- [ ] I have made corresponding changes to the documentation
- [x] My changes generate no new warnings
- [ ] I have added tests that prove my fix is effective or that my feature works
- [x] New and existing unit tests pass locally with my changes
- [x] Any dependent changes have been merged and published in downstream modules
- [x] Add a note to ``CHANGELOG.md``, summarizing the changes

aburrell commented on pysat/instruments/pysat_testing2d_xarray.py:
```diff
@@ -134,6 +134,11 @@ def load(fnames, tag=None, sat_id=None, malformed_index=False):
     latitude = 90.0 * np.cos(angle)
     data['latitude'] = (('time'), latitude)
 
+    # create constant altitude at 400 km
+    alt0 = 400.0
+    altitude = alt0 * np.ones(data['latitude'].shape)
+    data['altitude'] = (('time'), altitude)
+
```
So this whole bit, why not just add a single value as altitude to the coordinates?  `data.coords['altitude'] = alt0`.

jklenzing replied:
The values in the coordinates do not refer to the satellite altitude, but to the altitude of a remote sensing measurement (hence the 2d data).  It could definitely be clearer (see below).  I'm open to naming convention suggestions.

aburrell replied:
I would use "profile_height", since this is the height from the instrument (I assume?) rather than the altitude from the surface of the earth.  I would still add altitude to the coordinates, since part of the point of xarray is to not assign a bunch of the same values that aren't changing.

aburrell commented on pysat/instruments/pysat_testing2d_xarray.py:
```diff
@@ -145,15 +150,15 @@ def load(fnames, tag=None, sat_id=None, malformed_index=False):
     num = len(data['uts'])
     data['profiles'] = \
-        (('time', 'altitude'),
+        (('time', 'altitude_profile'),
```
Is this whole bit needed?  Altitude is constant.

jklenzing replied:
These are for the profile data, which becomes an issue for xarray.  The data profiles simulate some sort of remote instrument (think COSMIC) which provides data as a function of altitude from a fixed-altitude platform.  I'm up for a different name here, but it needs to be unique so xarray will work.

jklenzing commented:
altitude is a variable of time as far as xarray is concerned.

aburrell commented:
@jklenzing I think you are wrong.  Let's talk about this offline.

jklenzing commented:
I think we just need to define requirements better.  I've added comments and renamed some things to match what is being simulated.  Current xarray structure is:
```
<xarray.Dataset>
Dimensions:            (profile_height: 15, time: 864, x: 17, y: 17, z: 15)
Coordinates:
  * time               (time) datetime64[ns] 2009-01-01 ... 2009-01-01T23:58:20
  * profile_height     (profile_height) int64 0 1 2 3 4 5 6 7 8 9 10 11 12 13 14
    profile_height2    (time, z) float64 0.0 1.0 2.0 3.0 ... 11.0 12.0 13.0 14.0
    image_lat          (time, x, y) float64 0.0 1.0 2.0 3.0 ... 14.0 15.0 16.0
    image_lon          (time, x, y) float64 0.0 1.0 2.0 3.0 ... 14.0 15.0 16.0
Dimensions without coordinates: x, y, z
Data variables:
    uts                (time) datetime64[ns] 2009-01-01 ... 2009-01-01T23:58:20
    mlt                (time) float64 0.0 0.4124 0.8247 ... 19.05 19.46 19.88
    slt                (time) float64 0.08247 0.4948 0.9072 ... 19.55 19.96
    longitude          (time) float64 0.0 5.769 11.54 ... 287.3 293.1 298.8
    latitude           (time) float64 90.0 89.48 87.91 ... 24.46 33.65 42.45
    altitude           (time) float64 400.0 400.0 400.0 ... 400.0 400.0 400.0
    dummy1             (time) int64 0 0 0 1 1 2 2 2 ... 16 17 17 18 18 19 19 19
    dummy2             (time) int64 0 0 0 1 1 1 2 2 ... 17 17 18 18 18 19 19 19
    dummy3             (time) float64 0.0 0.0 0.0 ... 1.902e+04 1.902e+04
    dummy4             (time) int64 0 100 200 300 ... 86000 86100 86200 86300
    profiles           (time, profile_height) float64 0.0 0.0 ... 1.902e+04
    variable_profiles  (time, z) float64 0.0 0.0 0.0 ... 1.902e+04 1.902e+04
    images             (time, x, y) float64 0.0 0.0 0.0 ... 1.902e+04 1.902e+04
```
For use in pysatModels, should we move latitude, longtitude, and altitude to coordinates?

aburrell commented on pysat/instruments/pysat_testing2d_xarray.py:
```diff
@@ -119,21 +119,26 @@ def load(fnames, tag=None, sat_id=None, malformed_index=False):
     data['slt'] = (('time'), slt)
 
-    # create a fake longitude, resets every 6240 seconds
+    # create a fake longitude of satellite, resets every 6240 seconds
```
"satellite longitude"

aburrell commented on pysat/instruments/pysat_testing2d_xarray.py:
```diff
@@ -119,21 +119,26 @@ def load(fnames, tag=None, sat_id=None, malformed_index=False):
     data['longitude'] = (('time'), longitude)
 
-    # create latitude signal for testing polar orbits
+    # create fake latitude of satellite for testing polar orbits
```
"satellite latitude"

aburrell commented on pysat/instruments/pysat_testing2d_xarray.py:
```diff
@@ -119,21 +119,26 @@ def load(fnames, tag=None, sat_id=None, malformed_index=False):
     latitude = 90.0 * np.cos(angle)
     data['latitude'] = (('time'), latitude)
 
+    # create constant altitude at 400 km
```
"for a satellite that has yet to experience orbital decay"

jklenzing commented:
Updated comments as suggested.  Tweaked the xarray structure a little:
```
<xarray.Dataset>
Dimensions:                  (profile_height: 15, time: 864, x: 17, y: 17, z: 15)
Coordinates:
  * time                     (time) datetime64[ns] 2009-01-01 ... 2009-01-01T23:58:20
  * profile_height           (profile_height) int64 0 1 2 3 4 ... 10 11 12 13 14
    variable_profile_height  (time, z) float64 0.0 1.0 2.0 ... 12.0 13.0 14.0
    image_lat                (time, x, y) float64 0.0 1.0 2.0 ... 14.0 15.0 16.0
    image_lon                (time, x, y) float64 0.0 1.0 2.0 ... 14.0 15.0 16.0
Dimensions without coordinates: x, y, z
Data variables:
    uts                      (time) datetime64[ns] 2009-01-01 ... 2009-01-01T23:58:20
    mlt                      (time) float64 0.0 0.4124 0.8247 ... 19.46 19.88
    slt                      (time) float64 0.08247 0.4948 ... 19.55 19.96
    longitude                (time) float64 0.0 5.769 11.54 ... 293.1 298.8
    latitude                 (time) float64 90.0 89.48 87.91 ... 33.65 42.45
    altitude                 (time) float64 400.0 400.0 400.0 ... 400.0 400.0
    dummy1                   (time) int64 0 0 0 1 1 2 2 ... 17 17 18 18 19 19 19
    dummy2                   (time) int64 0 0 0 1 1 1 2 ... 17 18 18 18 19 19 19
    dummy3                   (time) float64 0.0 0.0 0.0 ... 1.902e+04 1.902e+04
    dummy4                   (time) int64 0 100 200 300 ... 86100 86200 86300
    profiles                 (time, profile_height) float64 0.0 ... 1.902e+04
    variable_profiles        (time, z) float64 0.0 0.0 ... 1.902e+04 1.902e+04
    images                   (time, x, y) float64 0.0 0.0 ... 1.902e+04
```

aburrell commented:
> For use in pysatModels, should we move latitude, longtitude, and altitude to coordinates?

No, it should be fine as is.  After all, this is multi-dimensional observations, not a model.  I was thinking in terms of this test instrument doubling as a ground-based platform.

jklenzing commented:
For consistency sake, we could treat the lat/lon/alt/slt/mlt as coordinates in both xarray objects.  Not sure how it works in analysis.

Should I include dummy coords for x,y,z?

aburrell commented:
> For consistency sake, we could treat the lat/lon/alt/slt/mlt as coordinates in both xarray objects. Not sure how it works in analysis.

Naw, let's leave it be.

> Should I include dummy coords for x,y,z?

Yes?  It would probably make it easier to plot things??

jklenzing commented:
done!
